### ISSUE TITLE
[NestedTensor] torch.compile silent specialization on _max_seqlen

### ISSUE THREAD

**Issue by tgalkovskyi** [pytorch/pytorch#124205]
As illustrated in code snippet below, NestedTensor._max_seqlen property is being silently specialized when compiled - problematic behavior (silent hardcoding of value used for tracing) can be reproduced in both `inductor` and `eager` compilation modes.

Expected behavior:
* compiler generated dynamic code that doesn't specialize into specific value of NestedTensor
OR at least fails prominently if specialization occurs - and guards kick in.

Observed behavior:
* success that leads to wrong outputs, as torch.compile produces specialized code without adding guards.

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer

```
from torch.nested._internal.nested_tensor import NestedTensor
import torch

@torch.compile
@torch._dynamo.config.patch(error_on_recompile=True)
def sample_fun(njt: NestedTensor) -> NestedTensor:
    njt = njt.clamp(0.1, 0.5)
    # if this is NOT specialized - this should lead to second printed output be larger than first
    njt *= njt._max_seqlen
    return njt

def create_njt(el_per_row):
    torch.manual_seed(0)
    njt = NestedTensor(
        values=torch.randn(10 * el_per_row, device="cuda"),
        offsets=torch.arange(11, device="cuda") * el_per_row,
    )
    # This sets _max_seqlen in cache
    print(njt._max_seqlen)
    return njt

with torch.inference_mode():
    # This works
    print(sample_fun(create_njt(el_per_row=1)).values())
    # But this printed output should be 2x as big - and it is NOT as max_seqlen is specialized in generated code.
    print(sample_fun(create_njt(el_per_row=2)).values())
```

When run with TORCH_LOGS=output_code this produces code that hardcodes _max_seqlen=1

https://gist.github.com/tgalkovskyi/691ccb841c318a795873bfa30c990b06

### Versions
ANP kernel pytorch (meta internal)

Version 2266
Built April 16, 2024, 5:22:27 AM

**Comment by jbschlosser** [comment]
I think if we properly guarded on subclass `ctx`, that would fix the correctness issue wrt recompiling. However, it's bad for perf to have regular recompiles as max seq len changes, so proper dynamic shapes support for this is needed.

Two possible directions:
* Add general dynamic shapes support for subclass attrs rather than treating them as constant (discussion with @bdhirsh indicates this isn't entirely straightforward and could take several weeks).
* Move the cache outside of the subclass instance as part of the nested int revamp. Min / max sequence length are immutable properties of a given `offsets` tensor, so it does make sense to track these alongside that. This approach solves the immediate problem of subclass attrs being treated as constant since the cached values would no longer be subclass attrs, but I think it's slightly TBD what this will look like wrt dynamic shapes when going this direction (cc @soulitzer for more context)

**Comment by jbschlosser** [comment]
Fixed in #122836, which adds dynamic shapes support for min / max sequence length by storing them as inner tensors of an NJT. Note that #129618 was also necessary for guard correctness.
